blockchain/src/pretty.rs

```diff
@@ -0,0 +1,405 @@
+    let aligned_key = format!("{:<kw$}", key, kw = key_width,);
+    let painted_key = match col {
+        Some(color) => color.paint(aligned_key),
+        None => aligned_key.into(),
+    };
+    write!(
+        f,
+        "{:>iw$}- {}:",
+        "",
+        painted_key,
+        iw = indent_size * indent_level,
+    )
+}
+
+fn fmt_val(
+    val: &Val,
+    f: &mut fmt::Formatter,
+    indent_size: usize,
+    indent_level: usize,
+    use_color: bool,
+) -> fmt::Result {
+    match val {
+        // write inline
+        Val::Single(_) => {
+            write!(f, " ")?;
+            fmt_pretty(val, f, indent_size, indent_level, use_color)?;
+            write!(f, "\n")
+        }
+        // write on the next line
+        Val::Pairs(_, _) => {
+            write!(f, "\n")?;
+            fmt_pretty(val, f, indent_size, indent_level, use_color)
+        }
+    }
+    // XXX: DRY up the duplicate calls to `fmt_pretty`?
+}
+
+fn fmt_pretty(
+    p: &Val,
+    f: &mut fmt::Formatter,
+    indent_size: usize,
+    indent_level: usize,
+    use_color: bool,
+) -> fmt::Result {
+    match p {
+        // format pretty-val as a terminal
+        Val::Single(display) => write!(f, "{}", display),
+        // format pretty-val as a set of  key-vals
+        Val::Pairs(color, pairs) => {
+            let key_width = longest_key_length(pairs);
+            pairs
+                .iter()
+                .fold(Ok(()), |prev_result, (key, val)| match prev_result {
+                    // return an error
+                    err @ Err(_) => err,
+                    // continue writing
+                    Ok(()) => {
+                        let col = color.and_then(|c| if use_color { Some(c) } else { None });
+                        fmt_key(key, f, indent_size, indent_level, col, key_width)?;
+                        fmt_val(val, f, indent_size, indent_level + 1, use_color)
+                    }
+                })
+        }
+    }
+}
+
+// implement display for a Val without color
+impl fmt::Display for Val {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt_pretty(
+            self,
+            f,
+            DISPLAY_INDENT_SIZE,
+            DISPLAY_INDENT_LEVEL,
+            DISPLAY_USE_COLOR,
+        )
+    }
+}
+
+// implement display for anything which implements Pretty
+impl fmt::Display for Pretty {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt_pretty(
+            &self.to_pretty(),
+            f,
+            DISPLAY_INDENT_SIZE,
+            DISPLAY_INDENT_LEVEL,
+            DISPLAY_USE_COLOR,
+        )
+    }
+}
+
+// format any Val to a string with color
+pub fn format_val(p: &Val, indent_size: usize) -> String {
+    // internal helper to get a formatter for calling `fmt_pretty`
+    struct Fmt<F>(F)
+    where
+        F: Fn(&mut fmt::Formatter) -> fmt::Result;
+
+    impl<F> fmt::Display for Fmt<F>
```

Comment:
The use of this local `Display` instance is suspicious. Unfortunately I don't have any suggestion to improve on this.

Reply:
it was necessary to get an instance of `fmt::Formatter` but @NicolasDP suggested an alternative structuring.. thanks!